libs/payments/cart/src/lib/cart.service.ts

```diff
@@ -93,18 +95,61 @@ export class CartService {
         errorReasonId = CartErrorReasonId.BASIC_ERROR;
       }
 
-      await this.cartManager
-        .finishErrorCart(cartId, {
+      // All unexpected/untyped errors should go to Sentry
+      if (errorReasonId === CartErrorReasonId.Unknown) {
+        Sentry.captureException(error, {
+          extra: {
+            cartId,
+          },
+        });
+      }
+
+      try {
+        await this.cartManager.finishErrorCart(cartId, {
           errorReasonId,
-        })
-        .catch((e) => {
-          // We silently fail here so as not to eat the original error
-          Sentry.captureException(e);
         });
 
-      // All unexpectd/untyped errors should go to Sentry
-      if (errorReasonId === CartErrorReasonId.Unknown) {
-        Sentry.captureException(error, {
+        const cart = await this.cartManager.fetchCartById(cartId);
+        if (cart.stripeSubscriptionId) {
+          const subscription = await this.subscriptionManager.retrieve(
+            cart.stripeSubscriptionId
+          );
+          const invoice = subscription.latest_invoice
+            ? await this.invoiceManager.retrieve(subscription.latest_invoice)
+            : undefined;
+          if (invoice) {
+            switch (invoice.status) {
+              case 'draft':
+                await this.invoiceManager.delete(invoice.id);
+                break;
+              case 'open':
+              case 'uncollectible':
+                await this.invoiceManager.void(invoice.id);
+                break;
+              case 'paid':
+                throw new CartError('Paid invoice found on failed cart', {
+                  cartId,
+                  stripeCustomerId: cart.stripeCustomerId,
+                  invoiceId: invoice.id,
+                });
+            }
+          }
+
+          const paymentIntent =
+            await this.subscriptionManager.getLatestPaymentIntent(subscription);
+          if (paymentIntent) {
+            await this.paymentIntentManager.cancel(paymentIntent.id);
```

Comment:
Should this be wrapped in a try/catch since we're expecting it to fail now and want to know if we cancel on a succeeded payment intent?

Reply:
I'm going to add an additional check to throw an error if the payment intent is in the "succeeded" state, and then wrap the cancellation in a try/catch that swallows the error. Any condition where the customer has paid shouldn't automatically lead to their subscription being cancelled. Its unlikely to come into effect either way though, due to the invoice-succeeded check above it